src/util/Scheduler.cpp

```diff
@@ -0,0 +1,266 @@
+        bool
+        shouldDrop(bool overloaded, Scheduler::AbsoluteDeadline now,
+                   Scheduler::Stats& stats) const
+        {
+            if (overloaded)
+            {
+                if (mDeadline != Scheduler::ABS_NEVER_DROP)
+                {
+                    stats.mActionsDroppedDueToOverload++;
+                    return true;
+                }
+            }
+            else
+            {
+                if (mDeadline != Scheduler::ABS_NEVER_DROP &&
+                    mDeadline != Scheduler::ABS_DROP_ONLY_UNDER_LOAD &&
+                    now > mDeadline)
+                {
+                    stats.mActionsDroppedDueToDeadline++;
+                    return true;
+                }
+            }
+            return false;
+        }
+    };
+
+    std::string mName;
+    nsecs mServiceTime{0};
+    std::deque<Element> mActions;
+
+  public:
+    Queue(std::string const& name) : mName(name)
+    {
+    }
+
+    std::string const&
+    name() const
+    {
+        return mName;
+    }
+
+    nsecs
+    serviceTime() const
+    {
+        return mServiceTime;
+    }
+
+    size_t
+    size() const
+    {
+        return mActions.size();
+    }
+
+    bool
+    isEmpty() const
+    {
+        return mActions.empty();
+    }
+
+    size_t
+    tryTrim(size_t loadLimit, Scheduler::AbsoluteDeadline now,
+            Scheduler::Stats& stats)
+    {
+        if (!mActions.empty())
+        {
+            bool overloaded = size() > loadLimit;
+            if (mActions.front().shouldDrop(overloaded, now, stats))
+            {
+                mActions.pop_front();
+                return 1;
+            }
+            if (mActions.back().shouldDrop(overloaded, now, stats))
+            {
+                mActions.pop_back();
+                return 1;
+            }
+        }
+        return 0;
+    }
+
+    void
+    enqueue(Action&& action, Scheduler::RelativeDeadline deadline)
+    {
+        auto elt = Element(std::move(action), deadline);
+        mActions.emplace_back(std::move(elt));
+    }
+
+    void
+    runNext(nsecs minServiceTime)
+    {
+        auto before = std::chrono::steady_clock::now();
+        Action action = std::move(mActions.front().mAction);
+        mActions.pop_front();
+        action();
+        auto after = std::chrono::steady_clock::now();
+        nsecs duration = std::chrono::duration_cast<nsecs>(after - before);
+        mServiceTime = std::max(mServiceTime + duration, minServiceTime);
+    }
+};
+
+Scheduler::Scheduler(size_t loadLimit,
+                     std::chrono::nanoseconds serviceTimeWindow)
+    : mQueueQueue(
+          [](std::shared_ptr<Queue> a, std::shared_ptr<Queue> b) -> bool {
+              return a->serviceTime() > b->serviceTime();
+          })
+    , mLoadLimit(loadLimit)
+    , mServiceTimeWindow(serviceTimeWindow)
+{
+}
+
+void
+Scheduler::trim(std::shared_ptr<Queue> q)
+{
+    Scheduler::AbsoluteDeadline now = std::chrono::steady_clock::now();
```

Comment:
Ok. This is a little delicate because of the difference between std::chrono::system_clock and std::chrono::steady_clock. We use system_clock in VirtualClock, but .. it's only appropriate for certain uses (eg. NTP-driven synchronization with the consensus time of the network), not others (eg. timer delays and scheduling, which should really be monotonic). I will take a look at cleaning this up, but it might get a little involved.

(I don't think we want to switch the entire system to monotonic time since it might drift out of consensus time with the network; maybe we can split the interfaces in VirtualClock between those that deal with consensus times and those that deal with monotonic times? Hmm..)

Reply:
Reworked VirtualClock::time_point to be steady_clock::time_point and use steady_clock::now, added VirtualClock::system_time_point and changed callers that need system times (eg. herder/SCP) to use it. Then switched Scheduler to use VirtualClock. This _seems_ to work.